File: src/lib/api/checkout.ts
/**
 * Dumb checkout client — productId + qty only; server owns prices.
 */

import { getApiBaseUrl } from "@/lib/api/config";

export type CheckoutLine = {
  productId: string;
  quantity: number;
};

export type CheckoutShipping = {
  fullName: string;
  line1: string;
  city: string;
  region: string;
  postal: string;
  country?: string;
};

export type CreateCheckoutSessionResult = {
  orderId: string;
  publicCode: string;
  clientSecret: string;
  totalCents: number;
  currency: "USD";
  status: string;
};

export type OrderItem = {
  productId: string;
  productSlug: string;
  productName: string;
  quantity: number;
  unitPriceCents: number;
  lineTotalCents: number;
  imageUrl: string;
};

export type OrderResult = {
  id: string;
  publicCode: string;
  email: string;
  status: string;
  currency: "USD";
  subtotalCents: number;
  shippingCents: number;
  totalCents: number;
  shippingAddress: Record<string, unknown>;
  items: OrderItem[];
  paidAt: string | null;
  createdAt: string;
};

export class ApiError extends Error {
  constructor(
    message: string,
    public code: string,
    public status: number,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

async function parseError(res: Response): Promise<ApiError> {
  let message = res.statusText || "Request failed";
  let code = "request_failed";
  try {
    const body = (await res.json()) as {
      detail?: { message?: string; code?: string } | string;
    };
    if (typeof body.detail === "string") {
      message = body.detail;
    } else if (body.detail && typeof body.detail === "object") {
      message = body.detail.message ?? message;
      code = body.detail.code ?? code;
    }
  } catch {
    /* ignore */
  }
  return new ApiError(message, code, res.status);
}

export async function createCheckoutSession(input: {
  email: string;
  lines: CheckoutLine[];
  shipping: CheckoutShipping;
}): Promise<CreateCheckoutSessionResult> {
  const res = await fetch(`${getApiBaseUrl()}/api/v1/checkout/sessions`, {
    method: "POST",
    headers: { "Content-Type": "application/json", Accept: "application/json" },
    body: JSON.stringify({
      email: input.email,
      lines: input.lines.map((l) => ({
        productId: l.productId,
        quantity: l.quantity,
      })),
      shipping: {
        fullName: input.shipping.fullName,
        line1: input.shipping.line1,
        city: input.shipping.city,
        region: input.shipping.region,
        postal: input.shipping.postal,
        country: input.shipping.country ?? "US",
      },
    }),
  });
  if (!res.ok) throw await parseError(res);
  return res.json() as Promise<CreateCheckoutSessionResult>;
}

export async function getOrder(
  orderId: string,
  email: string,
  options?: { sync?: boolean },
): Promise<OrderResult> {
  const url = new URL(`${getApiBaseUrl()}/api/v1/orders/${orderId}`);
  url.searchParams.set("email", email);
  if (options?.sync) {
    url.searchParams.set("sync", "true");
  }
  const res = await fetch(url.toString(), {
    method: "GET",
    headers: { Accept: "application/json" },
    cache: "no-store",
  });
  if (!res.ok) throw await parseError(res);
  return res.json() as Promise<OrderResult>;
}

File: backend/app/api/v1/schemas/checkout.py
"""Checkout / order API schemas — camelCase for FE."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class CamelModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        ser_json_by_alias=True,
    )


class CheckoutLineIn(CamelModel):
    product_id: str = Field(alias="productId")
    quantity: int = Field(ge=1, le=9)


class ShippingIn(CamelModel):
    full_name: str = Field(alias="fullName", min_length=1, max_length=255)
    line1: str = Field(min_length=1, max_length=255)
    city: str = Field(min_length=1, max_length=128)
    # UK county optional — deep validation in application layer
    region: str = Field(default="", max_length=64)
    postal: str = Field(min_length=1, max_length=32)
    country: str = Field(default="US", min_length=2, max_length=2)


class CreateCheckoutSessionIn(CamelModel):
    email: EmailStr
    lines: list[CheckoutLineIn] = Field(min_length=1)
    shipping: ShippingIn


class CreateCheckoutSessionOut(CamelModel):
    order_id: str = Field(serialization_alias="orderId")
    public_code: str = Field(serialization_alias="publicCode")
    client_secret: str = Field(serialization_alias="clientSecret")
    total_cents: int = Field(serialization_alias="totalCents")
    currency: Literal["USD"] = "USD"
    status: str


class OrderItemOut(CamelModel):
    product_id: str = Field(serialization_alias="productId")
    product_slug: str = Field(serialization_alias="productSlug")
    product_name: str = Field(serialization_alias="productName")
    quantity: int
    unit_price_cents: int = Field(serialization_alias="unitPriceCents")
    line_total_cents: int = Field(serialization_alias="lineTotalCents")
    image_url: str = Field(serialization_alias="imageUrl")


class OrderOut(CamelModel):
    id: str
    public_code: str = Field(serialization_alias="publicCode")
    email: str
    status: str
    currency: Literal["USD"] = "USD"
    subtotal_cents: int = Field(serialization_alias="subtotalCents")
    shipping_cents: int = Field(serialization_alias="shippingCents")
    total_cents: int = Field(serialization_alias="totalCents")
    shipping_address: dict[str, Any] = Field(serialization_alias="shippingAddress")
    items: list[OrderItemOut]
    paid_at: str | None = Field(default=None, serialization_alias="paidAt")
    created_at: str = Field(serialization_alias="createdAt")
